libraries/AP_Scripting/examples/DroneCAN_test.lua

```diff
@@ -0,0 +1,261 @@
+        byte_offset[1] = byte_offset[1] + 4
+        return ret
+    else
+        return nil
+    end
+end
+
+--[[
+    parse a parameter GetSet Value
+--]]
+local function parse_param_Value(payload, byte_offset)
+    local vtype = string.unpack("B", payload, byte_offset[1])
+    byte_offset[1] = byte_offset[1] + 1
+    if vtype == 0 then
+        return nil
+    elseif vtype == 1 then
+        -- integer (treat as 32 bit for now, actually 64 bit)
+        local ret = string.unpack("i", payload, byte_offset[1])
+        byte_offset[1] = byte_offset[1] + 8
+        return ret
+    elseif vtype == 2 then
+        -- float32
+        local ret = string.unpack("f", payload, byte_offset[1])
+        byte_offset[1] = byte_offset[1] + 4
+        return ret
+    elseif vtype == 3 then
+        -- bool
+        local v = string.unpack("B", payload, byte_offset[1])
+        byte_offset[1] = byte_offset[1] + 1
+        return v == 1
+    elseif vtype == 4 then
+        -- string
+        local slen = string.unpack("B", payload, byte_offset[1])
+        local ret = string.sub(payload, byte_offset[1]+1, slen+2)
+        byte_offset[1] = byte_offset[1] + 1 + slen
+        return ret
+    else
+        return nil
+    end
+end
+
+
+--[[
+    parse a parameter GetSet reply
+--]]
+local function parse_param_reply(payload)
+    local byte_offset = {1}
+    local value = parse_param_Value(payload, byte_offset)
+    local default_value = parse_param_Value(payload, byte_offset)
+    local max_value = parse_param_NumericValue(payload, byte_offset)
+    local min_value = parse_param_NumericValue(payload, byte_offset)
+    local name = string.sub(payload, byte_offset[1], #payload)
+    return name, value, default_value, min_value, max_value
+end
+
+local next_param_index = 0
+
+--[[
+    encode a 16 bit number as a DroneCAN int13
+--]]
+local function encode_int13(v)
+    return (v & 0xFF) | (v&0xFF00)<<3
+end
+
+local function fetch_param()
+    local payload, nodeid = param_handle:check_message()
+    if payload then
+        local pname, pvalue = parse_param_reply(payload)
+        if not pname or not pvalue then
+            gcs:send_text(MAV_SEVERITY.INFO, string.format("param restart loop %u", next_param_index))
+            next_param_index = 0
+        else
+            gcs:send_text(MAV_SEVERITY.INFO, string.format("got param reply from %u idx=%u '%s' : %f", nodeid, next_param_index, pname, pvalue))
+            next_param_index = next_param_index + 1
+        end
+    end
+    param_handle:request(TARGET_NODE:get(), string.pack("H",encode_int13(next_param_index)))
+end
+
+--[[
+    check for new NodeStatus messages
+--]]
+local function check_node_status()
+    local payload, nodeid = nodestatus_handle:check_message()
+    if not payload then
+        return
+    end
+    local uptime_sec, bits, _ = string.unpack("IBH", payload)
+    local health = bits&3
+    local mode = (bits>>2)&7
+    local sub_mode = (bits>>5)&7
```

Comment:
I thought that the helper would do all the random size bits for us? I thought the unpack would be of the `uavcan_protocol_NodeStatus` structure it seems like its the "wire" structure?

Reply:
it is the payload structure, it ties in at the message level, it doesn't use the compiled DSDL